File: implementations/java/provekit-lift-java-source/src/main/java/com/provekit/lift/java_source/RpcServer.java
package com.provekit.lift.java_source;

import com.provekit.ir.Jcs;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.List;

public final class RpcServer {
    public static final String Surface = "java";
    public static final String Dialect = "java-source";
    public static final String Version = "0.1.0-draft";

    private RpcServer() {}

    public static void run() throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out, true);
        String line;
        while ((line = in.readLine()) != null) {
            if (line.isBlank()) continue;
            Jcs.Obj response;
            try {
                response = handle((Jcs.Obj) Jcs.parse(line));
            } catch (RuntimeException e) {
                response = error(Jcs.nullValue(), -32700, "PARSE_ERROR: " + e.getMessage());
            }
            out.println(Jcs.encode(response));
        }
    }

    public static Jcs.Obj handle(Jcs.Obj request) {
        Jcs.Json id = request.get("id") == null ? Jcs.nullValue() : request.get("id");
        String method = request.stringFieldOrNull("method");
        if (method == null) return error(id, -32600, "INVALID_REQUEST: missing method");
        return switch (method) {
            case "initialize" -> initialize(id);
            case "lift" -> lift(id, request.get("params"));
            case "compile" -> compile(id, request.get("params"));
            case "shutdown" -> Jcs.object("jsonrpc", Jcs.string("2.0"), "id", id, "result", Jcs.nullValue());
            default -> error(id, -32601, "METHOD_NOT_FOUND: " + method);
        };
    }

    private static Jcs.Obj initialize(Jcs.Json id) {
        return Jcs.object(
            "jsonrpc", Jcs.string("2.0"),
            "id", id,
            "result", Jcs.object(
                "name", Jcs.string("provekit-lift-java-source"),
                "version", Jcs.string(Version),
                "protocol_version", Jcs.string("provekit-lift/1"),
                "capabilities", Jcs.object(
                    "authoring_surfaces", Jcs.array(Jcs.string(Dialect)),
                    "dialect", Jcs.string(Dialect),
                    "ir_version", Jcs.string("v1.1.0"),
                    "emits_signed_mementos", Jcs.bool(false)
                )
            )
        );
    }

    private static Jcs.Obj lift(Jcs.Json id, Jcs.Json paramsJson) {
        if (!(paramsJson instanceof Jcs.Obj params)) return error(id, -32602, "params required");
        String surface = params.stringFieldOrNull("surface");
        if (surface == null) surface = Dialect;
        if (!surface.equals(Surface) && !surface.equals(Dialect)) return error(id, 1003, "SURFACE_NOT_SUPPORTED: " + surface);
        Jcs.Json pathsJson = params.get("source_paths");
        if (!(pathsJson instanceof Jcs.Arr pathsArr) || pathsArr.isEmpty()) return error(id, -32602, "source_paths must be a non-empty array");
        List<String> paths = pathsArr.values().stream().map(v -> ((Jcs.Str) v).value()).toList();
        String root = params.stringFieldOrNull("workspace_root");
        if (root == null) root = ".";
        JavaSourceLifter.LiftResult result = new JavaSourceLifter().liftPaths(root, paths);
        return Jcs.object("jsonrpc", Jcs.string("2.0"), "id", id, "result", result.toJson());
    }

    private static Jcs.Obj compile(Jcs.Json id, Jcs.Json paramsJson) {
        if (!(paramsJson instanceof Jcs.Obj params)) return error(id, -32602, "params required");
        Jcs.Json ir = params.get("ir");
        if (ir == null) return error(id, -32602, "ir required");
        String body = new JavaSourceCompiler().compile(ir);
        return Jcs.object(
            "jsonrpc", Jcs.string("2.0"),
            "id", id,
            "result", Jcs.object("kind", Jcs.string("compiled-formula"), "body", Jcs.string(body))
        );
    }

    private static Jcs.Obj error(Jcs.Json id, int code, String message) {
        return Jcs.object(
            "jsonrpc", Jcs.string("2.0"),
            "id", id,
            "error", Jcs.object("code", Jcs.integer(code), "message", Jcs.string(message))
        );
    }
}

File: implementations/csharp/Provekit.Lift.Csharp/RpcServer.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Provekit.Lift.Csharp;

public static class RpcServer
{
    public const string Surface = "csharp";
    public const string Dialect = "csharp-source";

    public static void Run()
    {
        var lifter = new CsharpLifter();
        var compiler = new CsharpCompiler();
        string? line;
        while ((line = Console.In.ReadLine()) is not null)
        {
            line = line.Trim();
            if (string.IsNullOrEmpty(line)) continue;

            JsonNode? response;
            try
            {
                var parsed = JsonNode.Parse(line);
                if (parsed is null)
                {
                    response = Error(null, -32700, "PARSE_ERROR: null request");
                    Console.Out.WriteLine(response.ToJsonString());
                    Console.Out.Flush();
                    continue;
                }
                var request = parsed.AsObject();
                response = Dispatch(request, lifter, compiler);
            }
            catch (JsonException ex)
            {
                response = Error(null, -32700, $"PARSE_ERROR: {ex.Message}");
            }
            catch (InvalidOperationException ex)
            {
                response = Error(null, -32600, $"INVALID_REQUEST: {ex.Message}");
            }

            if (response is not null)
            {
                Console.Out.WriteLine(response.ToJsonString());
                Console.Out.Flush();
            }
        }
    }

    private static JsonObject Dispatch(JsonObject request, CsharpLifter lifter, CsharpCompiler compiler)
    {
        var id = request["id"];
        var methodNode = request["method"];
        var method = methodNode is not null ? methodNode.GetValue<string>() : "";

        return method switch
        {
            "initialize" => Initialize(id),
            "lift" => LiftRpc(id, request["params"] as JsonObject, lifter),
            "compile" => CompileRpc(id, request["params"] as JsonObject, compiler),
            "provekit.plugin.recognize" => RecognizeRpc(id, request["params"] as JsonObject),
            "shutdown" => new JsonObject { ["jsonrpc"] = "2.0", ["id"] = id?.DeepClone(), ["result"] = null },
            _ => Error(id, -32601, $"METHOD_NOT_FOUND: {method}"),
        };
    }

    public static JsonObject DispatchForTest(JsonObject request) =>
        Dispatch(request, new CsharpLifter(), new CsharpCompiler());

    private static JsonObject Initialize(JsonNode? id)
    {
        return new JsonObject
        {
            ["jsonrpc"] = "2.0",
            ["id"] = id?.DeepClone(),
            ["result"] = new JsonObject
            {
                ["name"] = "provekit-lift-csharp",
                ["version"] = "0.1.0",
                ["protocol_version"] = "pep/1.7.0",
                ["capabilities"] = new JsonObject
                {
                    ["authoring_surfaces"] = new JsonArray { Surface, Dialect },
                    ["ir_version"] = "v1.1.0",
                    ["emits_signed_mementos"] = false,
                },
            },
        };
    }

    private static JsonObject LiftRpc(JsonNode? id, JsonObject? paramsObj, CsharpLifter lifter)
    {
        if (paramsObj is null)
            return Error(id, -32602, "params required");

        var surface = paramsObj["surface"]?.GetValue<string>() ?? Surface;
        if (surface != Surface && surface != Dialect)
            return Error(id, 1003, $"SURFACE_NOT_SUPPORTED: {surface}");

        var sourcePathsNode = paramsObj["source_paths"];
        JsonArray? sourcePaths;
        try { sourcePaths = sourcePathsNode?.AsArray(); } catch { sourcePaths = null; }
        if (sourcePaths is null || sourcePaths.Count == 0)
            return Error(id, -32602, "source_paths must be a non-empty array of strings");

        var workspaceRoot = paramsObj["workspace_root"]?.GetValue<string>() ?? ".";

        var paths = sourcePaths.Select(p => p?.GetValue<string>() ?? "").Where(p => !string.IsNullOrEmpty(p)).ToList();
        if (paths.Count == 0)
            return Error(id, -32602, "source_paths must be a non-empty array of strings");

        try
        {
            var result = lifter.LiftPaths(workspaceRoot, paths);
            return LiftSuccessResponse(id, result);
        }
        catch (Exception ex)
        {
            return Error(id, -32603, ex.Message);
        }
    }

    private static JsonObject CompileRpc(JsonNode? id, JsonObject? paramsObj, CsharpCompiler compiler)
    {
        if (paramsObj is null)
            return Error(id, -32602, "params required");

        if (!paramsObj.ContainsKey("ir") || paramsObj["ir"] is null)
            return Error(id, -32602, "ir required");
        var ir = paramsObj["ir"];

        try
        {
            var result = compiler.Compile(ir);
            return new JsonObject
            {
                ["jsonrpc"] = "2.0",
                ["id"] = id?.DeepClone(),
                ["result"] = new JsonObject
                {
                    ["kind"] = "compiled-formula",
                    ["body"] = result,
                },
            };
        }
        catch (Exception ex)
        {
            return Error(id, -32603, ex.Message);
        }
    }

    private static JsonObject RecognizeRpc(JsonNode? id, JsonObject? paramsObj)
    {
        if (paramsObj is null)
            return Error(id, -32602, "params required");

        var sourcePathsNode = paramsObj["source_paths"];
        JsonArray? sourcePaths;
        try { sourcePaths = sourcePathsNode?.AsArray(); } catch { sourcePaths = null; }
        if (sourcePaths is null || sourcePaths.Count == 0)
            return Error(id, -32602, "source_paths must be a non-empty array of strings");

        var projectRoot = paramsObj["project_root"]?.GetValue<string>() ?? ".";
        var paths = sourcePaths.Select(p => p?.GetValue<string>() ?? "")
            .Where(p => !string.IsNullOrEmpty(p))
            .ToList();
        if (paths.Count == 0)
            return Error(id, -32602, "source_paths must be a non-empty array of strings");

        try
        {
            var directTemplates = BindingTemplatesFromParams(paramsObj);
            var templates = directTemplates.Count > 0
                ? directTemplates
                : CsharpBindingTemplateResolver.ResolveFromProject(projectRoot);
            var result = CsharpRecognizer.RecognizePaths(projectRoot, paths, templates);
            return new JsonObject
            {
                ["jsonrpc"] = "2.0",
                ["id"] = id?.DeepClone(),
                ["result"] = result,
            };
        }
        catch (Exception ex)
        {
            return Error(id, -32603, ex.Message);
        }
    }

    private static IReadOnlyList<JsonObject> BindingTemplatesFromParams(JsonObject paramsObj)
    {
        if (paramsObj["binding_templates"] is not JsonArray directTemplates)
            return Array.Empty<JsonObject>();

        var templates = new List<JsonObject>();
        foreach (var template in directTemplates)
        {
            if (template is JsonObject obj) templates.Add(obj);
        }

        return templates;
    }

    private static JsonObject LiftSuccessResponse(JsonNode? id, LiftResult result)
    {
        return new JsonObject
        {
            ["jsonrpc"] = "2.0",
            ["id"] = id?.DeepClone(),
            ["result"] = new JsonObject
            {
                ["kind"] = "ir-document",
                ["ir"] = JsonSerializer.SerializeToNode(result.Declarations),
                ["callEdges"] = JsonSerializer.SerializeToNode(result.CallEdges),
                ["diagnostics"] = JsonSerializer.SerializeToNode(result.Diagnostics),
                ["opacityReport"] = JsonSerializer.SerializeToNode(result.OpacityReport),
                ["refusals"] = JsonSerializer.SerializeToNode(result.Refusals),
            },
        };
    }

    private static JsonObject Error(JsonNode? id, int code, string message)
    {
        return new JsonObject
        {
            ["jsonrpc"] = "2.0",
            ["id"] = id?.DeepClone(),
            ["error"] = new JsonObject
            {
                ["code"] = code,
                ["message"] = message,
            },
        };
    }
}
